File: service/GoodNight.Service.Domain/Write/Quality.cs
using System.Collections.Immutable;
using GoodNight.Service.Domain.Write.Expressions;

namespace GoodNight.Service.Domain.Write
{
  public abstract record Quality(
    string Name,
    Type Type,
    string Raw,
    bool Hidden,
    string? Scene,
    string Description)
  {
    public string Urlname
    {
      get
      {
        return NameConverter.OfString(Name);
      }
    }


    public record Bool(
      string Name,
      string Raw,
      bool Hidden,
      string? Scene,
      string Description)
      : Quality(Name, Type.Bool, Raw, Hidden, Scene, Description) {}

    public record Int(
      string Name,
      string Raw,
      bool Hidden,
      string? Scene,
      string Description,
      int Minimum,
      int Maximum)
      : Quality(Name, Type.Int, Raw, Hidden, Scene, Description) {}

    public record Enum(
      string Name,
      string Raw,
      bool Hidden,
      string? Scene,
      string Description,
      IImmutableDictionary<int, string> Levels)
      : Quality(Name, Type.Enum, Raw, Hidden, Scene, Description) {}
  }
}

File: client/src/model/write/Quality.ts


export enum QualityType {
  Int = 1,
  Bool = 2,
  Enum = 3
}


// todo: consider different types of qualities in interface.

export interface Quality {
  name: string;
  urlname: string;

  raw: string;

  type: QualityType;
  description: string;
  hidden: boolean;
  scene: string | null;

  tags: Array<string>;
  category: Array<string>;

  // maximum: number | null;
  // minimum: number | null;
}
